src/randomizer.py

```diff
@@ -865,6 +865,186 @@ def minus1(node: ca.Node) -> ca.BinaryOp:
+            return randomize_associative_binop(node.name, node.subscript)
+        return ca.BinaryOp('+', node.name, node.subscript)
+
+    def deref(node: ca.Node) -> ca.UnaryOp:
+        """Surround the given node with a dereference operator"""
+        # This should be possible but it would mess up my crappy list handling code :|
+        #
+        #if isinstance(node, ca.UnaryOp) and node.op == '&':
+        #    return node.expr
+        return ca.UnaryOp('*', node)
+
+    def addr(node: ca.Node) -> ca.UnaryOp:
+        """Surround the given node with an address-of operator"""
+        # This should be possible but it would mess up my crappy list handling code :|
+        #
+        #if isinstance(node, ca.UnaryOp) and node.op == '*':
+        #    return node.expr
+        return ca.UnaryOp('&', node)
+
+    def rec(node: ca.Node, nodes: List[ca.Node]) -> Tuple[bool, bool]:
+        """Add each child node of the StructRef to a list, stopping when the node is not a
+        dereference or address-of unary operator. Returns 'failed' when a UnaryOp with an
+        op other than * or & was encountered."""
+        if isinstance(node, ca.UnaryOp):
+            if not node.op in ['&', '*']:
+                return False, True
+            else:
+                binop_or_array_ref, failed = rec(node.expr, nodes)
+                if failed:
+                    return binop_or_array_ref, failed
+                nodes.append(node)
+                return binop_or_array_ref, False
+        else:
+            nodes.append(node)
+            return isinstance(node, (ca.ArrayRef, ca.BinaryOp)), False
+
+        return False, True
+
+    # (Oh god why)
+    def apply_child(parent: ca.Node, func) -> None:
+        if isinstance(parent, ca.StructRef):
+            parent.name = func(parent.name)
+        elif isinstance(parent, ca.UnaryOp):
+            parent.expr = func(parent.expr)
+
+    def set_child(parent: ca.Node, child: ca.Node) -> None:
+        if isinstance(parent, ca.StructRef):
+            parent.name = child
+        elif isinstance(parent, ca.UnaryOp):
+            parent.expr = child
+
+    def get_child(parent: ca.Node) -> ca.Node:
+        if isinstance(parent, ca.StructRef):
+            return parent.name
+        elif isinstance(parent, ca.UnaryOp):
+            return parent.expr
+
+    struct_ref = random.choice(cands)
+    nodes: List[ca.Node] = []
+    binop_or_array_ref, failed = rec(struct_ref.name, nodes)
+    if failed:
+        #print('\033[91mPerm will cancel here\033[m')
+        return False
+
+    #cg = c_generator.CGenerator()
+    #print('\033[94m')
+    #print(cg.visit(struct_ref),end='')
+    #print('\033[m')
+
+    # Step 1: Simplify (...)->c to (*(...)).c
+    if struct_ref.type == '->':
+        struct_ref.type = '.'
+        struct_ref.name = deref(struct_ref.name)
+        nodes.append(struct_ref.name)
+        #print('-> changed: \033[32m',cg.visit(struct_ref),'\033[m')
+    nodes.append(struct_ref)
```

Comment:
At this point and forward I completely lose track of what's going on. Why do we recurse arbitrarily deep but only use indices [0], [1] and [2]? Can we rewrite the recursion as a loop? Can we give [0], [1], [2] names? types? explain their nesting structure? either this needs more comments, or preferably rewriting to avoid the list thing.